File: apps/demo-be-ts-effect/tests/integration/steps/token-lifecycle.steps.ts
import { Given, When, Then } from "@cucumber/cucumber";
import assert from "node:assert/strict";
import type { CustomWorld } from "../world.js";
import { TEST_JWT_SECRET } from "../hooks.js";
import * as jose from "jose";

// ---- Given: expired refresh token ----

Given("alice's refresh token has expired", async function (this: CustomWorld) {
  // Sign a refresh token with very short expiry (already expired)
  const secretKey = new TextEncoder().encode(TEST_JWT_SECRET);
  const expiredToken = await new jose.SignJWT({
    jti: `expired-${Date.now()}`,
    tokenType: "refresh",
  })
    .setProtectedHeader({ alg: "HS256" })
    .setSubject(this.userIds.get("alice") ?? "alice-id")
    .setIssuedAt(Math.floor(Date.now() / 1000) - 3600)
    .setExpirationTime(Math.floor(Date.now() / 1000) - 1800)
    .sign(secretKey);
  this.tokens.set("alice_refresh", expiredToken);
});

Given("alice has used her refresh token to get a new token pair", async function (this: CustomWorld) {
  const refreshToken = this.tokens.get("alice_refresh") ?? "";
  this.context["alice_original_refresh"] = refreshToken;
  const res = await this.post("/api/v1/auth/refresh", { refresh_token: refreshToken });
  if (res.status !== 200) {
    throw new Error(`Failed to refresh: ${JSON.stringify(res.body)}`);
  }
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  this.tokens.set("alice_access", (res.body as any).access_token as string);
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  this.tokens.set("alice_refresh", (res.body as any).refresh_token as string);
});

Given(/^the user "([^"]*)" has been deactivated$/, async function (this: CustomWorld, username: string) {
  const token = this.tokens.get(`${username}_access`) ?? "";
  const res = await this.post("/api/v1/users/me/deactivate", {}, token);
  if (res.status !== 200) {
    throw new Error(`Failed to deactivate ${username}: ${JSON.stringify(res.body)}`);
  }
});

Given("alice has already logged out once", async function (this: CustomWorld) {
  const token = this.tokens.get("alice_access") ?? "";
  const res = await this.post("/api/v1/auth/logout", {}, token);
  if (res.status !== 200) {
    throw new Error(`Logout failed: ${JSON.stringify(res.body)}`);
  }
});

// ---- When ----

When(/^alice sends POST \/api\/v1\/auth\/refresh with her refresh token$/, async function (this: CustomWorld) {
  const refreshToken = this.tokens.get("alice_refresh") ?? "";
  this.response = await this.post("/api/v1/auth/refresh", { refresh_token: refreshToken });
});

When(/^alice sends POST \/api\/v1\/auth\/refresh with her original refresh token$/, async function (this: CustomWorld) {
  const originalRefresh = this.context["alice_original_refresh"] as string;
  this.response = await this.post("/api/v1/auth/refresh", { refresh_token: originalRefresh });
});

When(/^alice sends POST \/api\/v1\/auth\/logout with her access token$/, async function (this: CustomWorld) {
  const token = this.tokens.get("alice_access") ?? "";
  this.response = await this.post("/api/v1/auth/logout", {}, token);
});

When(/^alice sends POST \/api\/v1\/auth\/logout-all with her access token$/, async function (this: CustomWorld) {
  const token = this.tokens.get("alice_access") ?? "";
  this.response = await this.post("/api/v1/auth/logout-all", {}, token);
});

// ---- Then ----

Then("the response body should contain an error message about token expiration", function (this: CustomWorld) {
  assert.ok(this.response !== null);
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const body = this.response?.body as Record<string, string>;
  const message = (body["message"] ?? body["error"] ?? "").toLowerCase();
  assert.ok(message.length > 0);
});

Then("the response body should contain an error message about invalid token", function (this: CustomWorld) {
  assert.ok(this.response !== null);
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const body = this.response?.body as Record<string, string>;
  const message = (body["message"] ?? body["error"] ?? "").toLowerCase();
  assert.ok(message.length > 0);
});

Then("alice's access token should be invalidated", async function (this: CustomWorld) {
  const token = this.tokens.get("alice_access") ?? "";
  const res = await this.get("/api/v1/users/me", token);
  assert.strictEqual(res.status, 401);
});

File: apps/demo-be-rust-axum/tests/integration/steps/token_lifecycle_steps.rs
use cucumber::{given, then, when};

use crate::world::AppWorld;

#[when("alice sends POST /api/v1/auth/refresh with her refresh token")]
async fn refresh_with_alice_token(world: &mut AppWorld) {
    let token = world.refresh_token.clone().unwrap_or_default();
    world.svc_refresh(&token).await;
    if world.last_status == 200 {
        world.auth_token = world
            .last_body
            .get("accessToken")
            .and_then(|v| v.as_str())
            .map(String::from);
        world.refresh_token = world
            .last_body
            .get("refreshToken")
            .and_then(|v| v.as_str())
            .map(String::from);
    }
}

#[given("alice's refresh token has expired")]
async fn alice_refresh_expired(world: &mut AppWorld) {
    // Use an obviously expired/invalid token
    world.refresh_token = Some(
        "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJ0ZXN0IiwiZXhwIjoxfQ.invalid".to_string(),
    );
}

#[given("alice has used her refresh token to get a new token pair")]
async fn alice_used_refresh_token(world: &mut AppWorld) {
    let original = world.refresh_token.clone().unwrap_or_default();
    world.original_refresh_token = Some(original.clone());

    world.svc_refresh(&original).await;
    if world.last_status == 200 {
        world.auth_token = world
            .last_body
            .get("accessToken")
            .and_then(|v| v.as_str())
            .map(String::from);
        world.refresh_token = world
            .last_body
            .get("refreshToken")
            .and_then(|v| v.as_str())
            .map(String::from);
    }
}

#[when("alice sends POST /api/v1/auth/refresh with her original refresh token")]
async fn refresh_with_original_token(world: &mut AppWorld) {
    let token = world.original_refresh_token.clone().unwrap_or_default();
    world.svc_refresh(&token).await;
}

#[given(expr = "the user {string} has been deactivated")]
async fn user_deactivated_step(world: &mut AppWorld, _username: String) {
    let token = world.auth_token.clone().unwrap_or_default();
    world.svc_deactivate(&format!("Bearer {token}")).await;
}

#[when("alice sends POST /api/v1/auth/logout with her access token")]
async fn logout_alice(world: &mut AppWorld) {
    let token = world.auth_token.clone().unwrap_or_default();
    world.svc_logout(&format!("Bearer {token}")).await;
}

#[when("alice sends POST /api/v1/auth/logout-all with her access token")]
async fn logout_all_alice(world: &mut AppWorld) {
    let token = world.auth_token.clone().unwrap_or_default();
    world.svc_logout_all(&format!("Bearer {token}")).await;
}

#[then("alice's access token should be invalidated")]
async fn check_token_invalidated(world: &mut AppWorld) {
    let token = world.auth_token.clone().unwrap_or_default();
    world.svc_get_profile(&format!("Bearer {token}")).await;
    assert_eq!(
        world.last_status, 401,
        "Expected 401 after logout, got {}",
        world.last_status
    );
}

#[given("alice has already logged out once")]
async fn alice_already_logged_out(world: &mut AppWorld) {
    let token = world.auth_token.clone().unwrap_or_default();
    world.svc_logout(&format!("Bearer {token}")).await;
}

#[then("the response body should contain an error message about token expiration")]
async fn error_token_expired(world: &mut AppWorld) {
    let msg = world
        .last_body
        .get("message")
        .and_then(|v| v.as_str())
        .unwrap_or("");
    assert!(
        !msg.is_empty(),
        "Expected token expiration error, got: {}",
        world.last_body
    );
}

#[then("the response body should contain an error message about invalid token")]
async fn error_invalid_token(world: &mut AppWorld) {
    let msg = world
        .last_body
        .get("message")
        .and_then(|v| v.as_str())
        .unwrap_or("");
    assert!(
        !msg.is_empty(),
        "Expected invalid token error, got: {}",
        world.last_body
    );
}
